internal/pkg/earpointer/taint.go

```diff
@@ -0,0 +1,191 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package earpointer
+
+import (
+	"go/types"
+
+	"github.com/google/go-flow-levee/internal/pkg/source"
+	"golang.org/x/tools/go/ssa"
+)
+
+// A data structure to assist the traversal of an EAR heap.
+type heapTraversal struct {
+	partitions *Partitions
+	callees    map[*ssa.Function]bool // the functions containing the references of interest
+	visited    ReferenceSet           // the visited references during the traversal
+}
+
+// For a function, transitively get the functions called within this function.
+// Argument "depth" controls the depth of the call chain.
+// For example, return {g1,g2,g3} for "func f(){ g1(); g2() }, func g1(){ g3() }".
+func getCalleeFunctions(fn *ssa.Function, result map[*ssa.Function]bool, depth uint) {
+	if depth <= 0 {
+		return
+	}
+	for _, b := range fn.Blocks {
+		for _, instr := range b.Instrs {
+			if call, ok := instr.(*ssa.Call); ok {
+				// TODO: the call graph can be CHA, RTA, VTA, etc.
+				if callee := call.Call.StaticCallee(); callee != nil && len(callee.Blocks) > 0 {
+					if _, ok := result[callee]; !ok {
+						result[callee] = true
+						getCalleeFunctions(callee, result, depth-1)
+					}
+				}
+			}
+		}
+	}
+}
+
+func GetCalleeFunctions(fn *ssa.Function, depth uint) map[*ssa.Function]bool {
+	result := make(map[*ssa.Function]bool)
+	result[fn] = true
+	getCalleeFunctions(fn, result, depth)
+	return result
+}
+
+func (ht *heapTraversal) isWithinFunctions(ref Reference) bool {
+	if fn := ref.Value().Parent(); fn != nil {
+		_, ok := ht.callees[fn]
+		return ok
+	}
+	// Globals and Builtins have no parents.
+	return true
+}
+
+// Obtain the references associated with a taint source, with field sensitivity.
+// Consider only the references located with the "callees".
+func GetSrcRefs(src *source.Source, isTaintField func(named *types.Named, index int) bool,
+	state *Partitions, callees map[*ssa.Function]bool) ReferenceSet {
+
+	val, ok := src.Node.(ssa.Value)
+	if !ok {
+		return nil
+	}
+	rep := state.Representative(MakeLocalWithEmptyContext(val))
+	refs := make(ReferenceSet)
+	ht := &heapTraversal{partitions: state, callees: callees, visited: make(ReferenceSet)}
+	ht.getStructFieldRefs(rep, val.Type(), refs, isTaintField)
+	return refs
+}
+
+// Obtain the references associated with a taint source, with field sensitivity.
+// Composite types are examined recursively to identify the taint elements, e.g.
+// (1) when a map contains taint elements, these elements are examined to identify
+//     taint sources; and
+// (2) when a struct object contains taint fields, only these fields
+// 	   and all their subfields are included in taint sources, and the struct object
+//     itself is a taint source. Other fields will not be tainted.
+func (ht *heapTraversal) getStructFieldRefs(rep Reference, tp types.Type, result ReferenceSet, isTaintField func(named *types.Named, index int) bool) {
```

Comment:
Push "fixed for a given instance" fields into heapTraversal. This is [likely] mildly more efficient. More importantly it is less code and less complicated signatures. `result` and `isTaintField` look like they fall into this.

Reply:
Now `isTaintField` has been moved into `heapTraversal`.
`result` is kept as it is the output of this function.